File: problems/07_trees/0235_lowest_common_ancestor_of_a_binary_search_tree/solution.py
"""
Problem: 0235. Lowest Common Ancestor of a BST
Difficulty: Medium
URL: https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-search-tree/

解法：利用 BST 排序性沿樹走，O(h)/O(1)。
"""

from __future__ import annotations

from typing import Optional, List

from collections import deque

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val; self.left = left; self.right = right

def from_level(arr: List[Optional[int]]) -> Optional[TreeNode]:
    if not arr: return None
    root = TreeNode(arr[0]); q = deque([root]); i = 1
    while q and i < len(arr):
        n = q.popleft()
        if i < len(arr) and arr[i] is not None: n.left = TreeNode(arr[i]); q.append(n.left)
        i += 1
        if i < len(arr) and arr[i] is not None: n.right = TreeNode(arr[i]); q.append(n.right)
        i += 1
    return root

def find(root, val):
    cur = root
    while cur:
        if cur.val == val: return cur
        cur = cur.left if val < cur.val else cur.right
    return None

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        pass


if __name__ == "__main__":
    solution = Solution()

    # Test 1
    # result = solution.solve(...)
    # assert result == expected

    print("All tests passed!")

File: problems/07_trees/0235_lowest_common_ancestor_of_a_binary_search_tree/solution.dart
// Problem: 0235. Lowest Common Ancestor of a BST
// Difficulty: Medium

class TreeNode {
  int val; TreeNode? left, right;
  TreeNode([this.val = 0, this.left, this.right]);
}

TreeNode? fromLevel(List<int?> a) {
  if (a.isEmpty) return null;
  final root = TreeNode(a[0]!);
  final q = <TreeNode>[root]; var qi = 0, i = 1;
  while (qi < q.length && i < a.length) {
    final n = q[qi++];
    if (i < a.length && a[i] != null) { n.left = TreeNode(a[i]!); q.add(n.left!); }
    i++;
    if (i < a.length && a[i] != null) { n.right = TreeNode(a[i]!); q.add(n.right!); }
    i++;
  }
  return root;
}

TreeNode? find(TreeNode? root, int v) {
  // TODO: implement
  return null;
}

TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
  // TODO: implement
  throw UnimplementedError();
}


void main() {
  // Test 1
  // final result = solve(...);
  print('All tests passed!');
}
